src/main/java/org/tikv/common/log/SlowLogSpanImpl.java

```diff
@@ -17,78 +17,76 @@
 
-import static org.tikv.common.log.SlowLogImpl.getSimpleDateFormat;
-
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
-import java.text.SimpleDateFormat;
+import java.util.HashMap;
+import java.util.Map;
 
 public class SlowLogSpanImpl implements SlowLogSpan {
   private final String name;
-  private final long requestStartNS;
-  private final long requestStartMS;
+  private final long requestStartInstantNS;
+  private final long requestStartUnixNS;
+
+  /** Key-Value pairs which will be logged, e.g. function name, key, region, etc. */
+  private final Map<String, String> properties;
 
-  private long startMS;
-  private long endMS;
   /**
    * use System.nanoTime() to calculate duration, cause System.currentTimeMillis() is not monotonic
    */
-  private long startNS;
+  private long startInstantNS;
 
-  private long endNS;
+  private long endInstantNS;
 
-  public SlowLogSpanImpl(String name, long requestStartMS, long requestStartNS) {
+  public SlowLogSpanImpl(String name, long requestStartMS, long requestStartInstantNS) {
     this.name = name;
-    this.requestStartMS = requestStartMS;
-    this.requestStartNS = requestStartNS;
-    this.startMS = 0;
-    this.startNS = 0;
-    this.endMS = 0;
-    this.endNS = 0;
+    this.requestStartUnixNS = requestStartMS * 1_000_000;
+    this.requestStartInstantNS = requestStartInstantNS;
+    this.properties = new HashMap<>();
+    this.startInstantNS = 0;
+    this.endInstantNS = 0;
+  }
+
+  @Override
+  public void addProperty(String key, String value) {
+    properties.put(key, value);
   }
 
   @Override
   public void start() {
-    startNS = System.nanoTime();
-    startMS = requestStartMS + (startNS - requestStartNS) / 1_000_000;
+    startInstantNS = System.nanoTime();
   }
 
   @Override
   public void end() {
-    endNS = System.nanoTime();
-    endMS = startMS + (endNS - startNS) / 1_000_000;
+    endInstantNS = System.nanoTime();
   }
 
   @Override
   public JsonElement toJsonElement() {
-    SimpleDateFormat simpleDateFormat = getSimpleDateFormat();
     JsonObject jsonObject = new JsonObject();
-    jsonObject.addProperty("name", name);
-    jsonObject.addProperty("start", getStartString(simpleDateFormat));
-    jsonObject.addProperty("end", getEndString(simpleDateFormat));
-    jsonObject.addProperty("duration", getDurationString());
+    jsonObject.addProperty("event", name);
+    jsonObject.addProperty("begin_unix_ns", getStartUnixNS());
+    jsonObject.addProperty("duration_ns", getDurationNS());
 
-    return jsonObject;
-  }
-
-  private String getStartString(SimpleDateFormat simpleDateFormat) {
-    if (startMS == 0) {
-      return "N/A";
+    if (!properties.isEmpty()) {
+      JsonObject propertiesObject = new JsonObject();
+      for (Map.Entry<String, String> entry : properties.entrySet()) {
+        propertiesObject.addProperty(entry.getKey(), entry.getValue());
+      }
+      jsonObject.add("properties", propertiesObject);
```

Comment:
should it be put into the above `if` statement?

Reply:
Which `if` statement? I think it is already in an if statement.

https://github.com/tikv/client-java/blob/f560809ff8ba78950811f0b61d22347da5644744/src/main/java/org/tikv/common/log/SlowLogSpanImpl.java#L71-L77